File: arrays/longest_consecutive_sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        numsSet = set(nums)

        maxSeqLen = 0 # minimum length

        for n in nums:
            # if n-1 is in the numsSet, then we do not iterate upwards
            # if n-1 is not in numsSet, then we check the sequence length
            # we simply walk each subsequence of numbers once

            if n-1 not in numsSet:
                seqLen = 1
                while n+1 in numsSet:
                    seqLen += 1
                    n += 1

                maxSeqLen = max(maxSeqLen, seqLen)

        return maxSeqLen

File: arrays/longest_consecutive_sequence.cpp
class Solution {
public:
  int longestConsecutive(vector<int> &nums) {
    std::set<int> num_set(nums.begin(), nums.end());
    int global_best{0};

    for (int i : num_set) {
      if (num_set.count(i - 1) != 1) {
        int local_best = 1;
        while (num_set.count(i + 1) == 1) {
          local_best += 1;
          i += 1;
        }
        global_best = std::max(global_best, local_best);
      }
    }
    return global_best;
  }
};
